opencog/rule-engine/forwardchainer/DefaultForwardChainerCB.cc

```diff
@@ -56,47 +60,58 @@ DefaultForwardChainerCB::DefaultForwardChainerCB(AtomSpace& as,
  */
 vector<Rule*> DefaultForwardChainerCB::choose_rules(FCMemory& fcmem)
 {
+    new_rules.clear(); //empty vector before starting matching
     Handle source = fcmem.get_cur_source();
     if (source == Handle::UNDEFINED)
         throw InvalidParamException(TRACE_INFO, "Needs a valid source atom");
 
     vector<Rule*> chosen_rules;
     auto rules = fcmem.get_rules();
+    vector<Rule*> visited;
     for (Rule* rule : rules) {
         HandleSeq impl_members = rule->get_implicant_seq();
         bool match = false;
-
         for (Handle h : impl_members) {
             //exceptions
-            if(h->getType() == ABSENT_LINK) continue;
+            if (h->getType() == ABSENT_LINK)
+                continue;
 
-            AtomSpace rule_atomspace;
-            Handle hcpy = rule_atomspace.add_atom(h);
-            Handle implicant_vardecl = rule_atomspace.add_atom(
+            AtomSpace temp_pm_as;
+            Handle hcpy = temp_pm_as.add_atom(h);
+            Handle implicant_vardecl = temp_pm_as.add_atom(
                     gen_sub_varlist(h, rule->get_vardecl()));
-            Handle sourcecpy = rule_atomspace.add_atom(source);
-
-            HandleSeq hx { implicant_vardecl, hcpy, hcpy };
-            BindLinkPtr bl = createBindLink(hx);
-            DefaultImplicator imp(&rule_atomspace);
-            imp.implicand = bl->get_implicand();
-            bl->imply(imp);
-
-            for (const auto& hi : imp.result_list) {
-                if (hi == sourcecpy) {
-                    match = true;
-                    break;
+            Handle sourcecpy = temp_pm_as.add_atom(source);
+
+            BindLinkPtr bl = createBindLink(HandleSeq { implicant_vardecl,
+                                                         hcpy,hcpy });
+            VarGroundingPMCB gcb(&temp_pm_as);
+            gcb.implicand = bl->get_implicand();
+            bl->imply(gcb);
+
+            FindAtoms fv(VARIABLE_NODE);
+            for (const auto& termg_map : gcb.term_groundings) {
+                for (const auto& it : termg_map) {
+                    if (it.second == sourcecpy) {
+                        match = true;
+                        fv.search_set(it.first);
+
+                        HandleSeq new_rules = substitute_rule_part(
+                                temp_pm_as,
+                                temp_pm_as.add_atom(rule->get_handle()),
+                                fv.varset, gcb.var_groundings);
+                        for (const auto& nr : new_rules)
+                            if (find(new_rules.begin(), new_rules.end(),
+                                     nr) == new_rules.end())
+                                new_rules.push_back(nr);
```

Comment:
I am trying to understand what this loop is doing...  

So, for each Handle `nr` in `new_rules`, check if `nr` is **not** in `new_rules` and add `nr` into `new_rules`...? I don't get it.